### PR TITLE
Stub Orders Gateway

### PR DESCRIPTION
**Pull request by jamesatheyDDS**
## Description

 * Adds stub implementations of the 4 Orders Gateway API endpoints.
 * Makes these new endpoints accessible only from the orders hostname.
 * Updates README.md to cover the Orders Gateway and its swagger API.
 * Adds the Orders swagger definition to the `bin/gen_server.sh` script and the `.pre-commit-config.yaml` check.

## Reviewer Notes

I did not change the app_detector to be aware of the orders hostname. Instead, the orders SubMux branches directly off of the `site` Mux. The Orders Gateway will use client certificates for authentication and is also sessionless, so it was not appropriate to branch off of the `root` SubMux.

Instead, the orders path has its own orders hostname detector, which just checks that the orders hostname matches and rejects everything else. As a result, none of the member, office, or tsp paths are reachable from the orders hostname, and none of the orders paths are reachable from the my, office, or tsp hostnames.

I also made the top-level `site` handle the static public content (e.g., `/downloads`, `/swagger-ui`, `favicon.ico`, etc.) directly, bypassing the session and hostname middleware that are part of the `root SubMux`. This makes the `/orders/v0/docs` path work correctly, while still only being accessible from the orders hostname.

## Code Review Verification Steps

* [x] All tests pass.
* [x] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/blob/master/docs/backend.md#logging)
* [x] The requirements listed in
 [Querying the Database Safely](https://github.com/transcom/mymove/blob/master/docs/backend.md#querying-the-database-safely)
 have been satisfied.
* [x] Any migrations/schema changes also update the diagram in docs/schema/dp3.sqs
* [x] There are no aXe warnings for UI.
* [x] This works in IE.
* Any new client dependencies (Google Analytics, hosted libraries, CDNs, etc) have been:
  * [x] Communicated to @willowbl00
  * [x] Added to the list of [network dependencies](https://github.com/transcom/mymove#client-network-dependencies)
* [x] Request review from a member of a different team.

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/154947024) for this change

### CODE REVIEW

File: cmd/webserver/main.go [hunk 1]
```diff
@@ -72,16 +72,18 @@ func httpsComplianceMiddleware(inner http.Handler) http.Handler {
...
 
-	build := flag.String("build", "build", "the directory to serve static files from.")
+	build := flag.String("build", "public", "the directory to serve static files from.")
 	config := flag.String("config-dir", "config", "The location of server config files")
 	env := flag.String("env", "development", "The environment to run in, which configures the database.")
 	listenInterface := flag.String("interface", "", "The interface spec to listen for connections on. Default is all.")
 	myHostname := flag.String("http_my_server_name", "localhost", "Hostname according to environment.")
 	officeHostname := flag.String("http_office_server_name", "officelocal", "Hostname according to environment.")
 	tspHostname := flag.String("http_tsp_server_name", "tsplocal", "Hostname according to environment.")
+	ordersHostname := flag.String("http_orders_server_name", "orderslocal", "Hostname according to environment.")
```
Comment (chrisgilmerproj) [inline cmd/webserver/main.go]: You'll need to update a value here: https://github.com/transcom/mymove/blob/master/config/app.container-definition.json#L53-L64

I made the same mistake recently in my own PR.  Without it you'll just get 400's on experimental, staging and prod.

File: cmd/webserver/main.go [hunk 2]
```diff
@@ -232,6 +234,21 @@ func main() {
...
+	site.Handle(pat.Get("/favicon.ico"), clientHandler)
+
+	ordersMux := goji.SubMux()
+	ordersDetectionMiddleware := auth.OrdersDetectorMiddleware(logger, *ordersHostname)
+	ordersMux.Use(ordersDetectionMiddleware)
+	ordersMux.Use(noCacheMiddleware)
+	ordersMux.Handle(pat.Get("/swagger.yaml"), fileHandler(*ordersSwagger))
+	ordersMux.Handle(pat.Get("/docs"), fileHandler(path.Join(*build, "swagger-ui", "orders.html")))
+	ordersMux.Handle(pat.New("/*"), handlers.NewOrdersAPIHandler(handlerContext))
+	site.Handle(pat.Get("/orders/v0/*"), ordersMux)
```
Comment (chrisgilmerproj) [inline cmd/webserver/main.go]: I'm not sure if we wanted to start with `v0` or `v1` for new apis.  I'd probably just say `v1` but maybe I don't have the history on this.

File: pkg/auth/orders_detector.go [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+import (
+	"net/http"
+	"strings"
+
+	"go.uber.org/zap"
+)
+
+// OrdersDetectorMiddleware only allows ordersHost through
+func OrdersDetectorMiddleware(logger *zap.Logger, ordersHostname string) func(next http.Handler) http.Handler {
```
Comment (chrisgilmerproj) [inline pkg/auth/orders_detector.go]: You can probably just use the `DetectorMiddleware` from `pkg/auth/app_detector.go` instead of writing a new one here.

File: pkg/handlers/orders_api.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
+package handlers
```
Comment (chrisgilmerproj) [inline pkg/handlers/orders_api.go]: It's fine to name this `pkg/handlers/orders.go` since I think all of the files in this directory are part of the API.

File: cmd/webserver/main.go (hunk 1 above)
Comment (jamesatheyDDS) [reply cmd/webserver/main.go]: Thanks for catching that. Fixed.

File: swagger/orders.yaml [hunk 5]
```diff
@@ -13,8 +13,6 @@ schemes:
 paths:
   /orders:
     get:
-      tags:
-        - orders
```
Comment (chrisgilmerproj) [inline swagger/orders.yaml]: Was there a reason you removed the tags? I'm just curious because I recently added them to the TSP public API.  I figured its a nice way to keep things organized but it also changes your namespacing for imports and the names of your functions.  So I guess either way is fine because this is still a tiny API.

File: pkg/auth/orders_detector.go (hunk 3 above)
Comment (jamesatheyDDS) [reply pkg/auth/orders_detector.go]: I went that route at first. I decided against it for several reasons:
1. When I wired the orders SubMux directly to the `site`, instead of branching off of `root`, I found that I couldn't reuse the same instance of the DetectorMiddleware in my SubMux without causing a crash. I could have instantiated another copy, I suppose.

Comment (jamesatheyDDS) [reply pkg/auth/orders_detector.go]: I went that route at first. I decided against it for several reasons:

1. When I wired the orders SubMux directly to the site, instead of branching off of root, I found that I couldn't reuse the same instance of the DetectorMiddleware in my SubMux without causing a crash. I could have instantiated another copy, I suppose.
1. I can blindly reject everything that's _not_ the orders hostname from /orders/* with my middleware, instead of adding additional branches in each handler to check if the request came from the orders hostname.
1. The Orders API is sessionless and stateless, and will use client certs instead of login.gov for authentication. If I use DetectorMiddleware, I have to look at the session variable to see if it's the right hostname. 
1. If DetectorMiddleware does _not_ know about the orders host, then all requests it sees from the orders host will immediately fail with "Bad hostname," blocking requests going to that hostname from the rest of the app by default. If it _does_ know about the orders host, I need to look at all of the other places where IsMyApp() or IsOfficeApp() is called and decide how to correctly reject the orders hostname.

File: pkg/handlers/orders_api.go (hunk 4 above)
Comment (jamesatheyDDS) [reply pkg/handlers/orders_api.go]: There's already a `pkg/handlers/orders.go` file, which contains the orders endpoints from the `internal` API. I can put all of my handlers in that file, but that would be mixing endpoints from two different swagger definitions.

File: swagger/orders.yaml (hunk 5 above)
Comment (jamesatheyDDS) [reply swagger/orders.yaml]: I removed the tags because:
1. I only have 1 tag,
1. it was making all of my generated method names longer
1. it was adding an unnecessary subdirectory, making my imports uglier

**Comment by jamesatheyDDS** [comment]
@chrisgilmerproj There's no make orders_client_run because there is no orders client. It's just an external server-to-server API.

File: cmd/webserver/main.go (hunk 2 above)
Comment (jamesatheyDDS) [reply cmd/webserver/main.go]: I'm not dogmatic about it. If `v1` is preferred at this point, I'll change it.

The API is not yet deployed, and although the definition has settled down quite a bit after many iterations with the users (USAF, USMC, USN), I am painfully aware of the possibility for the API to change somewhat drastically. (There's an impedance mismatch between the Orders API's idea of amendments and the existing DB schema, for example.)

File: pkg/handlers/orders_api.go (hunk 4 above)
Comment (chrisgilmerproj) [reply pkg/handlers/orders_api.go]: Aha, so what I ended up doing for TSP in a recent (as of today) PR was putting stuff in the same file and then adding big comment sections and prefixing my TSP api handlers with the word `Public`.  You could probably also use the prefix `Public` if you wanted.  

A different tact may be that we start separating out the files into separate directories by API type?  Since you're struggling with this and I'm also struggling with this maybe we need better separation.

For now probably just leave it as `orders_api.go` and we'll figure it out later.

File: swagger/orders.yaml (hunk 5 above)
Comment (chrisgilmerproj) [reply swagger/orders.yaml]: All very good points.

**Comment by chrisgilmerproj** [comment]
Ahh, I hadn't really thought through that you didn't have a client.  So you can scratch the `Makefile` comment.  

Your CircleCI tests may be failing because CircleCI seems to be having some kind of outage right now.  Probably will come back online in a while.

**Comment by akostibas** [comment]
Code generally looks fine, though I'm curious what the local development story looks like. I tried making some requests (`curl http://orderslocal:3000`) and wound up getting `Invalid Host header`.

It's also not clear to me how to get Swagger to serve up its yaml file:
screen shot 2018-07-19 at 4 22 13 pm

File: cmd/webserver/main.go [hunk 6]
```diff
@@ -72,16 +72,18 @@ func httpsComplianceMiddleware(inner http.Handler) http.Handler {
 
 func main() {
 
-	build := flag.String("build", "build", "the directory to serve static files from.")
+	build := flag.String("build", "public", "the directory to serve static files from.")
```
Comment (chrisgilmerproj) [inline cmd/webserver/main.go]: Was this change intentional? I think this would change a lot of stuff in our depoyments.

Comment (jamesatheyDDS) [reply cmd/webserver/main.go]: Yes, it was intentional. I actually thought this was a mistake in the code, because there is no "build" folder in the git clone, just a "public" one. I can easily take this out, and just set the flag locally so that it works in my checkout.

Comment (chrisgilmerproj) [reply cmd/webserver/main.go]: The `build` directly is created by the `Makefile`  and not checked in.  This is where all the compiled stuff ends up going.  I'd revert this change.

As for why its not working for you, there's some code in the Makefile that manages this stuff when you have a client.  You don't have a client and thus you didn't get the cool default scripts.  For instance you have to run `make client_build` to get the `build/` directory.  It does `yarn install && bin/copy_swagger_ui.sh` for you and puts everything in the right place.  I get this by default for the TSP client because I have this in the `Makefile`:

```
tsp_client_run: client_deps
  HOST=tsplocal yarn start
```
